syntax = "proto3";

// message StatusRequest {
//     string query = 1;
//     int16 status = 2;  it is a mandatory field
//     int16 date = 3;
//     repeated string list_strings = 4;  repeated: the field might be repeated from 0 - infinite number of times
//     optional int32 random = 5;   it is an optional field
// }

// Service for obtaining list of S&P500 companies
service ListSearchSP{
    // Obtainig list of S&P500 companies
    rpc ListCompaniesSP(SearchRequestSP) returns (SearchResponseSP) {}

    // Streaming the information
    rpc ListCompaniesSPStreamSingleResponse(stream SearchRequestSP) returns (SearchResponseSP) {}
    rpc ListCompaniesSPRequestStreamResponse (SearchRequestSP) returns (stream SearchResponseSP) {}
    rpc ListCompaniesSPStreamRequestStreamResponse (stream SearchRequestSP) returns (stream SearchResponseSP) {}
}

// We ask for the available companies that there are on S&P500
message SearchRequestSP{
    string query = 1;
}

message Company{
    // Name & Ticker of the company
    string name = 1;
    string ticker = 2;

    // Which is the sector of the company
    message Sector{
        string principal = 1;
        string secondary = 2;
    }

    // The date of the first time added to the listing of S&P500
    string first_added = 3;

    // The year in which it was founded
    int32 year_foundation = 4;
}

// Obtaining a list of all the names of the companies that are actually on the S&P500
message SearchResponseSP{
    repeated Company = 1;
}
